File: algorithms/cpp/606.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    string helper(TreeNode *node) {
        string s = "";
        if (node == NULL) return s;
        s += to_string(node->val);
        if (node->left != NULL || node->right != NULL) s += "(" + helper(node->left) + ")";
        if (node->right != NULL) s += "(" + helper(node->right) + ")";
        return s;
    }
public:
    string tree2str(TreeNode *t) {
        return helper(t);
    }
};

int main() {
    return 0;
}

File: algorithms/rust/606.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    #[allow(dead_code)]
    pub fn tree2str(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        if let Some(root) = root {
            let left = Self::tree2str(root.borrow().left.clone());
            let right = Self::tree2str(root.borrow().right.clone());
            if !right.is_empty() {
                format!("{}({})({})", root.borrow().val, left, right)
            } else if !left.is_empty() {
                format!("{}({})", root.borrow().val, left)
            } else {
                format!("{}", root.borrow().val)
            }
        } else {
            "".to_string()
        }
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
